### PR TITLE
Add partition value check before writing

### PR DESCRIPTION
**Pull request by shixuan-fan**
Add this check to fail early if it encounters invalid partition values.

### CODE REVIEW

**Comment by shixuan-fan** [comment]
Kind ping @electrum @haozhun

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 1]
```diff
@@ -474,7 +476,11 @@ private void validateSchema(Optional<String> partitionName, Properties schema)
                 partitionValues.add(HIVE_DEFAULT_DYNAMIC_PARTITION);
             }
             else {
-                partitionValues.add(value.toString());
+                String valueString = value.toString();
+                if (!CharMatcher.inRange((char) 0x20, (char) 0x7E).matchesAllOf(valueString)) {
+                    throw new PrestoException(HIVE_INVALID_PARTITION_VALUE, "\"" + valueString + "\" contains characters not in the range of 0x20 - 0x7E");
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: Error message need to mention partition key and Hive. Users don't generally see error code. And it is important that the error message provides user with enough information.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedQueries.java [hunk 2]
```diff
@@ -106,6 +106,15 @@ public void testRcTextCharDecoding()
         testRcTextCharDecoding(true);
     }
 
+    @Test
+    public void testInvalidPartitionValue()
+            throws Exception
+    {
+        assertUpdate("CREATE TABLE invalid_partition_value (a int, b varchar) WITH (partitioned_by = ARRAY['b'])");
+        assertQueryFails("INSERT INTO invalid_partition_value VALUES (4, '\n')", "\"\n\" contains characters not in the range of 0x20 - 0x7E");
+        assertUpdate("DROP TABLE invalid_partition_value");
```
Comment (haozhun) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedQueries.java]: You can do 'create table as select' instead. This way, you won't need drop. And you will only need one query.

Alternatively, you could also test both CTAS and INSERT. It's up to you.

**Comment by shixuan-fan** [comment]
Changed error message and added one more test with CTAS.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java [hunk 3]
```diff
@@ -474,7 +476,11 @@ private void validateSchema(Optional<String> partitionName, Properties schema)
                 partitionValues.add(HIVE_DEFAULT_DYNAMIC_PARTITION);
             }
             else {
-                partitionValues.add(value.toString());
+                String valueString = value.toString();
+                if (!CharMatcher.inRange((char) 0x20, (char) 0x7E).matchesAllOf(valueString)) {
+                    throw new PrestoException(HIVE_INVALID_PARTITION_VALUE, "Partition key \"" + valueString + "\" in Hive contains characters not in the range of 0x20 - 0x7E");
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java]: Cannot use "valueString" as value of Hive partition key. Partition columns in Hive can only contain characters in the range of 0x20 - 0x7E.

**Pull request merged**